src/services/pcn-bridge/src/Bridge.cpp

```diff
@@ -1,143 +1,342 @@
+    if (it->name() == port.name()) {
+      continue;
+    }
+    logger()->trace("Sending packet to port {0}", it->name());
+    it->send_packet_out(packet, tagged, vlan);
+  }
 }
 
-// Basic default implementation, place your extension here (if needed)
 std::shared_ptr<Ports> Bridge::getPorts(const std::string &name) {
-  // call default implementation in base class
   return BridgeBase::getPorts(name);
 }
 
-// Basic default implementation, place your extension here (if needed)
 std::vector<std::shared_ptr<Ports>> Bridge::getPortsList() {
-  // call default implementation in base class
   return BridgeBase::getPortsList();
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::addPorts(const std::string &name, const PortsJsonObject &conf) {
+  std::lock_guard<std::mutex> guard(ports_mutex_);
+
   BridgeBase::addPorts(name, conf);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::addPortsList(const std::vector<PortsJsonObject> &conf) {
-  // call default implementation in base class
   BridgeBase::addPortsList(conf);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::replacePorts(const std::string &name,
                           const PortsJsonObject &conf) {
-  // call default implementation in base class
   BridgeBase::replacePorts(name, conf);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::delPorts(const std::string &name) {
-  // call default implementation in base class
+  std::lock_guard<std::mutex> guard(ports_mutex_);
+
   BridgeBase::delPorts(name);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::delPortsList() {
-  // call default implementation in base class
   BridgeBase::delPortsList();
 }
 
 std::shared_ptr<Fdb> Bridge::getFdb() {
-  throw std::runtime_error("Bridge::getFdb: Method not implemented");
+  if (fdb_ == nullptr || fdb_.use_count() == 0)
+    throw std::runtime_error("Fdb does not exists");
+
+  return fdb_;
 }
 
 void Bridge::addFdb(const FdbJsonObject &value) {
-  throw std::runtime_error("Bridge::addFdb: Method not implemented");
+  if (fdb_ != nullptr)
+    throw std::runtime_error("Fdb already exists");
+
+  fdb_ = std::make_shared<Fdb>(*this, value);
 }
 
-// Basic default implementation, place your extension here (if needed)
 void Bridge::replaceFdb(const FdbJsonObject &conf) {
-  // call default implementation in base class
-  BridgeBase::replaceFdb(conf);
+  delFdb();
+
+  if (conf.agingTimeIsSet()) {
+    fdb_->setAgingTime(conf.getAgingTime());
+  }
 }
 
 void Bridge::delFdb() {
-  throw std::runtime_error("Bridge::delFdb: method not implemented");
+  // don't remove fdb. simply flush it.
+  if (fdb_ == nullptr) {
+    // this should never happen
+    throw std::runtime_error("There is not filtering database in the bridge");
+  }
+
+  fdb_->flush();
+  fdb_->setAgingTime(300);
```

Comment:
I am not sure this is ok. I am confused a little bit about the meaning of the create and delete operations on a container, special in this one that is always present. 
Maybe we can just forbid the creation and deletion of Fdb, anyway, this is always going to be present.

Reply:
I got the idea from the simplebridge service. But i agree with you, it would be better if we forbid the manual creation/deletion of Fdb.